avidemux_core/ADM_coreImage/src/ADM_imageOperation.cpp

```diff
@@ -60,7 +60,10 @@ bool ADMImage::duplicateMacro(ADMImage *src,bool swap)
                         }
                     }
                 }
-                BitBlit(dest, destStride,source,sourceStride,opWidth, opHeight);
+                if (destStride == sourceStride)
+                    memcpy(dest, source, destStride*opHeight);  // copying in one run is faster then line-by-line
```

Comment:
If a single memcpy call wins over an SSE4-accelerated implementation in libavtuil (it certainly does on non-x86, so ARM users should rejoyce), wouldn't it be appropriate to handle this special case in BitBlit itself rather than letting callers deal with it individually?

Reply:
there is NO SSE4 accelerated stuff in libavutil

Reply:
BitBlit can not use a single memcpy, becuase it doesn't know important parameters

Reply:
> there is NO SSE4 accelerated stuff in libavutil

Purely my reading glitch, sorry.

> BitBlit can not use a single memcpy, becuase it doesn't know important parameters

If the caller uses memory beyond the width of the picture but before the end of the stride for some purposes, the operation will be destructive, yes. Is the assumption safe that this cannot happen in `ADMImage::duplicateMacro`?

En-bloc copy evaluates only that source and destination strides match. Have I missed something else?

Reply:
BitBlit is used in other scenarios, similar to a "Picture-in-picture" function, where a full copy would destroy visible pixels.
In ADMImage::duplicateMacro the width and height match is asserted, only the stride remains, which is checked before memcpy. I dont see a possibility (copying same size ADM_images) where the BitBlit would be used instead of memcpy, but left there just in case.

Reply:
Ah, indeed, in `ADMImage::copyTo`, thank you for pointing out.